### PR TITLE
[ABD] use fletcher_4 routines natively with `abd_iterate_func()`

### PR DESCRIPTION
**Pull request by dpquigl**
- export ABD compatible interface from fletcher_4

- add ABD fletcher_4 tests for data and metadata ABD types.

Reviewed-by: David Quigley <david.quigley@intel.com>
Rebased-by: David Quigley<david.quigley@intel.com>
Signed-off-by: Gvozden Neskovic <neskovic@gmail.com>

### CODE REVIEW

**Comment by behlendorf** [comment]
@dpquigl please rebase this on the latest master to resolve the kernel.org build failures.

File: cmd/ztest/ztest.c [hunk 1]
```diff
@@ -5644,18 +5644,26 @@ ztest_fletcher(ztest_ds_t *zd, uint64_t id)
 	while (gethrtime() <= end) {
 		int run_count = 100;
 		void *buf;
+		struct abd *abd_data, *abd_meta;
 		uint32_t size;
+
```
Comment (behlendorf) [inline cmd/ztest/ztest.c]: [cstyle] We don't want extra white space in the variable declarations.  Please remove.

File: cmd/ztest/ztest.c [hunk 2]
```diff
@@ -5671,9 +5679,30 @@ ztest_fletcher(ztest_ds_t *zd, uint64_t id)
 			VERIFY0(bcmp(&zc, &zc_ref, sizeof (zc)));
 			VERIFY0(bcmp(&zc_byteswap, &zc_ref_byteswap,
 			    sizeof (zc_byteswap)));
+
+			/* Test ABD - data */
+
```
Comment (behlendorf) [inline cmd/ztest/ztest.c]: [cstyle] Additional white space should be dropped.

File: cmd/ztest/ztest.c [hunk 3]
```diff
@@ -5671,9 +5679,30 @@ ztest_fletcher(ztest_ds_t *zd, uint64_t id)
...
+			abd_fletcher_4_byteswap(abd_data, size, NULL, &zc_byteswap);
+			abd_fletcher_4_native(abd_data, size, NULL, &zc);
+
+			VERIFY0(bcmp(&zc, &zc_ref, sizeof (zc)));
+			VERIFY0(bcmp(&zc_byteswap, &zc_ref_byteswap,
+			    sizeof (zc_byteswap)));
+
+			/* Test ABD - metadata */
+
```
Comment (behlendorf) [inline cmd/ztest/ztest.c]: [cstyle] Additional white space should be dropped.

File: module/zcommon/zfs_fletcher.c [hunk 4]
```diff
@@ -435,7 +436,7 @@ fletcher_4_impl_get(void)
 		break;
 	}
 
-	ASSERT3P(ops, !=, NULL);
+	VERIFY(ops);
```
Comment (behlendorf) [inline module/zcommon/zfs_fletcher.c]: Let's leave this as an ASSERT3P.  There's no compelling reason to promote it to a VERIFY.

File: module/zcommon/zfs_fletcher.c [hunk 5]
```diff
@@ -778,6 +779,92 @@ fletcher_4_fini(void)
...
 
+/* ABD adapters */
+
+static void
+abd_fletcher_4_init(zio_abd_checksum_data_t *cdp)
+{
+	const fletcher_4_ops_t *ops = fletcher_4_impl_get();
+	cdp->acd_private = (void *) ops;
+
+	ASSERT(ops);
```
Comment (behlendorf) [inline module/zcommon/zfs_fletcher.c]: Let's drop this ASSERT, we've just called `fletcher_4_impl_get()` which has an identical check.

File: module/zcommon/zfs_fletcher.c [hunk 6]
```diff
@@ -778,6 +779,92 @@ fletcher_4_fini(void)
...
+
+static void
+abd_fletcher_4_fini(zio_abd_checksum_data_t *cdp)
+{
+	fletcher_4_ops_t *ops = (fletcher_4_ops_t *) cdp->acd_private;
+
+	ASSERT(ops);
+
+	if (cdp->acd_byteorder == ZIO_CHECKSUM_NATIVE) {
+		if (ops->fini_native)
```
Comment (behlendorf) [inline module/zcommon/zfs_fletcher.c]: Neither `fini_native` or `fini_byteswap` are optional.  It's safe to call them w/out the conditional.

File: module/zcommon/zfs_fletcher.c [hunk 7]
```diff
@@ -829,4 +916,6 @@ EXPORT_SYMBOL(fletcher_4_native_varsize);
 EXPORT_SYMBOL(fletcher_4_byteswap);
 EXPORT_SYMBOL(fletcher_4_incremental_native);
 EXPORT_SYMBOL(fletcher_4_incremental_byteswap);
+
+EXPORT_SYMBOL(fletcher_4_abd_ops);
```
Comment (behlendorf) [inline module/zcommon/zfs_fletcher.c]: There shouldn't be any need to make this symbol visible outside the zcommon module.  So this can be removed.

**Comment by behlendorf** [comment]
@dpquigl when you get a chance you can refresh this and address the review comments.

**Comment by dpquigl** [comment]
Applied the review comments and rebased against master. Hopefully for the last time.

**Pull request merged**